Added SQL soak test for IMap as a data source

PavloBida opened the pull request.

devOpsHazelcast commented:
[![CLA assistant check](https://cla.hazelcast.org/pull/badge/not_signed)](https://cla.hazelcast.org/hazelcast/hazelcast-jet-ansible-tests?pullRequest=131) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.hazelcast.org/hazelcast/hazelcast-jet-ansible-tests?pullRequest=131) before we can accept your contribution.<br/><hr/>**Pavlo Bida** seems not to be a GitHub user. You need a GitHub account to be able to sign the CLA. If you have already a GitHub account, please [add the email address used for this commit to your account](https://help.github.com/articles/why-are-my-commits-linked-to-the-wrong-user/#commits-are-not-linked-to-any-user).<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.hazelcast.org/check/hazelcast/hazelcast-jet-ansible-tests?pullRequest=131) it.</sub>

PavloBida commented:
@olukas , please review.
I addressed your comments and made the following changes:

- removed 'soak' from test namings
- added timeout between queries, which can be configured with 'queryTimeout' run param
- data size can now be configured with 'dataSetSize' run param
- changed IMap name to 'object_sql_map'
- added assertion that verifies that sql query has returned at least one row
- added logging, which will print execution percentage (interval between prints depends on test duration, it will print progress update every durationInMillis / 100 milliseconds). Please let me know if you like it or not
- added assertion that verifies that sql queries are being executed. Will fail the test if no queries executed in durationInMillis / 100 time interval
- added executable 'main' method
- other minor fixes

olukas commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+import com.hazelcast.map.IMap;
+import com.hazelcast.sql.SqlResult;
+import com.hazelcast.sql.SqlService;
+import org.junit.Assert;
```
Rather use assert methods from `AbstractSoakTest`, otherwise you need to put junit on server classpath.

PavloBida replied:
Done

olukas reviewed (changes requested):
Added some comments. There's also a lot of checkstyle rules violations which need to be fixed (run `mvn clean install -DskipTests` from `sql-imap-test` module).

olukas commented:
@gurbuzali We can start to use Jet soak test environment for soak testing SQL (I think we can start with that once we will use Platform for Jet soak tests env). 

However we're currently using `jet submit` for running Jet test. I guess it's not appropriate for SQL (even if it should probably work), I think we should provide some ability to rather use directly HZ client for tests which are not related to Jet. It can be implemented in `AbstractSoakTest`.

We should probably also have different profiles for running:
- stream processing tests (i.e. former Jet tests)
- Platform tests which are not related to stream processing (we can use this testsuite also for tests like this)
- SQL tests
- all tests

Generally we can adjust this testsuite to be usable for whole Platform. Some work (and probably not trivial) will be needed on `hazelcast-jet-ansible` side as well (no worry, it can be done by me or anyone from SDET team).

It's not urgent now (we should currently focus on Jet 4.5) but I would like to know your opinion about it.

PavloBida commented on pom.xml:
```diff
@@ -41,6 +41,7 @@
         <module>grpc-test</module>
         <module>sort-test</module>
         <module>file-ingestion-test</module>
+        <module>sql-imap-test</module>
```
Done

PavloBida commented on sql-imap-test/pom.xml:
```diff
@@ -0,0 +1,106 @@
+        <dependency>
+            <groupId>com.hazelcast</groupId>
+            <artifactId>hazelcast-sql</artifactId>
+            <version>4.2</version>
```
Done

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+            }
+
+            //Timeout between queries to not stress out the cluster
+            sleep(queryTimeout);
```
Done

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+    private long onePercentInMillis;
+    private long currentIteration = 0;
+    private long currentQueryCount = 0;
+    private long lastQueryCount = 0;
```
Oof, my bad, thanks

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/SqlObjectMapTest.java:
```diff
@@ -0,0 +1,52 @@
+
+    private static final String MAP_NAME = "object_sql_map";
+
+    private HazelcastInstance hazelcastInstance;
```
Why?

olukas replied:
It's not an issue but I don't like it since it bypasses the structure from `AbstractSoakTest`. On the other hand I don't have a technical argument for not to have it like this since it's just a holder => feel free to keep it as it is.

Btw. should not it be init in `AbstractSqlMapTest` instead of classes which extends it?

PavloBida replied:
Ok, I changed that to JetInstance. I was passing it to one of the methods within AbstractSqlMapTest, so maybe it does make sense to make it work with JetInstance rather than HazelcastInstance. Thanks!

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+        for (long i = 0; i < DEFAULT_DATA_SET_SIZE; i++) {
+            map.put(new Key(i), new Pojo(i));
+        }
+        assertEquals("Failed to populate the map", DEFAULT_DATA_SET_SIZE, map.size());
```
Done

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+    }
+
+    private boolean isQuerySuccessful(SqlResult sqlResult) {
+        return sqlResult.iterator().hasNext();
```
Would such kind of assertion be really in scope of this test? 
We have that covered in functional tests.

olukas replied:
I think we should have some evidence that the results are not completely incorrect. Currently we check we just get "some result". I don't think we should check every item in the results but I think the check for length and 1 of item is useful to be sure that it's not completely broken.

Sure, we have similar tests in functional tests but how do we know that something doesn't go wrong after 30hours etc.? We should have some simple evidence that the result can be considered as correct.

PavloBida replied:
I guess that makes sense!
Added field value check.

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/SqlObjectMapTest.java:
```diff
@@ -0,0 +1,52 @@
+    @Override
+    protected void init(JetInstance client){
+        hazelcastInstance = client.getHazelcastInstance();
+        dataSetSize = propertyInt("dataSetSize", DEFAULT_DATA_SET_SIZE);
```
Done

PavloBida commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+    private void printProgress() {
+        long timer = System.currentTimeMillis() - begin;
+        long nextPrintTime = onePercentInMillis * currentIteration;
+        boolean toPrint = timer >= nextPrintTime;
```
Ok, I'll make it print progress per certain period of time.
But which period to you suggest? 1 minute, or 5 or more?

olukas replied:
I would keep it as simple as possible - just log it after some amount of queries is executed.

PavloBida replied:
Ok, so I made it to print the progress each 500 queries (that's about 1 minute, considering our sql throughput).
That will result us with about ~3000 lines printed for 48h run. @olukas please let me know if you consider that reasonable.
Also I added time elapsed indicator to the print message. Now the output looks like this:
```
13:46:43.254 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Running...
13:46:56.937 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 0m, 13s. Executed 100 queries
13:47:08.526 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 0m, 25s. Executed 200 queries
13:47:19.930 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 0m, 36s. Executed 300 queries
13:47:31.217 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 0m, 47s. Executed 400 queries
13:47:42.426 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 0m, 59s. Executed 500 queries
13:47:53.625 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 1m, 10s. Executed 600 queries
13:48:04.746 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 1m, 21s. Executed 700 queries
13:48:15.886 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 1m, 32s. Executed 800 queries
13:48:27.032 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 1m, 43s. Executed 900 queries
13:48:38.145 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Time elapsed: 0d, 0h, 1m, 54s. Executed 1000 queries
13:48:43.269 [ INFO] [c.h.j.t.s.t.SqlBinaryMapTest] Test completed successfully. Executed 1046 queries in 0d, 0h, 2m, 0s.
```

olukas replied:
Looks good, thanks

olukas commented on sql-imap-test/src/main/java/com/hazelcast/jet/tests/sql/tests/AbstractSqlTest.java:
```diff
@@ -0,0 +1,156 @@
+
+    private void assertNotStuck() {
+        Assert.assertNotEquals(
+                String.format("No queries executed in %d seconds.", MILLISECONDS.toSeconds(durationInMillis)),
```
This is still an issue - the message is incorrect, it still says "No queries executed in durationInMillis seconds." where `durationInMillis` is the whole tests duration.

olukas reviewed (commented):
Added couple of comments. There are still a lot of Checkstyle rules violations. 

I didn't check the new tests since we should adjust `hazelcast-jet-ansible` to be able to run those tests. I try to prioritize it and hopefully start to work on it during this week.

PavloBida commented:
@olukas all the review comments have been resolved. Checkstyle errors fixed.
Please have a look.

PavloBida commented:
Pull request closed because of CLA issues
